### DISCUSSION TITLE
3 (!) HTTP servers for NextCloud installation?!!

### DISCUSSION THREAD

**Discussion by DimanNe** [nextcloud/all-in-one#4394] [Questions] [answered]
I have just started looking into NextCloud docker containers/images (=> I might be missing something), but if I count correctly, this official guide ([Reverse Proxy Documentation](https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md)) recommends using 3 (!!!) HTTP servers:

* `apache` in nextcloud_nextcloud-aio-apache_1
* `caddy` in nextcloud_nextcloud-aio-apache_1
* another one, like nginx.

Assuming I would really like to have my own reverse proxy (like Cloudflare's new `pingora` **written in Rust**, or nginx), why do I have to use `apache` and `caddy`?

Actually, I have same question for the case if I do not want to have my own reverse proxy: why are there two servers?
Why is one not enough?

P.S. Also, it appears that `APACHE_PORT` name is misleading/wrong, as Caddy is using it https://github.com/nextcloud/all-in-one/blob/main/Containers/apache/Caddyfile#L18

**Comment by Zoey2936** [comment, marked as answer]
Apache connects to php and handels nextclound
Caddy handels https and prixues to all other services (push, colabora, etc.)
And both still exist in reverse proxy mode to make configuration simpler

**Reply by DimanNe** [reply]
Yeah, I understand this more or less... I think Caddy is used for ACME/SSL things, and apache translates http to fastcgi.
And I also need another one in front of these two.

But I still think that final result is sub-optimal, because we end up with 3 http servers => 3x latency, 3x CPU/MEM consumption, 3x larger attack surface, 3x more configuration (and room for mis-configuration, even though it is "hidden" from end-users), etc....

**Reply by Zoey2936** [reply]
I know, but nextcloud is made to use the .htaccess of apache, but apache doesn't provide easy https and you also don't want a complicated nginx config for example..

**Reply by DimanNe** [reply]
@Zoey2936 Are there any **obvious** downsides/issues if I get rid of Caddy and proxy from nginx/pingora directly to apache?
(assuming managing certificates myself is fine)

**Reply by Zoey2936** [reply]
I think you can't connect to apache from outside of the dockercontainer and you would need to fully recreate the caddyfile. Also you might miss changes of the caddyfile

**Reply by docjyJ** [reply]
Would it be possible to separate the Apache server from the Caddy server?

I use another Caddy server to multiplex other services than Nextcloud. So I don't need Nextcloud AIO to be HTTPS. And even, I don't need to open the Apache server because I can access it via the AIO network. This would be great to make this project more customizable.

Solution idea: move the Apache server into the Nextcloud container.
- The Nextcloud container would become the place where the Nextcloud application is executed.
- Apache would become the proxy container, places to reach the different services.
- Other containers for each service, Postgresql, Redis, etc.

**Reply by Zoey2936** [reply]
1. aio forces https
2. apache currently readonly mode, which doesn't work if apache runs in the nextcloud container
3. there is no real advantage to not use caddy in the chain of proxies

**Reply by docjyJ** [reply]
I understand.
It would be nice to have just one caddy proxy instead of two (in my case).

**Reply by docjyJ** [reply]
I just found this: https://github.com/nextcloud/all-in-one/pull/3192
